runtime/near-wallet-contract/implementation/wallet-contract/src/internal.rs

```diff
@@ -35,32 +36,95 @@ pub fn parse_rlp_tx_to_action(
     target: &AccountId,
     context: &ExecutionContext,
     expected_nonce: &mut u64,
-) -> Result<(near_action::Action, TransactionValidationOutcome), Error> {
+) -> Result<(near_action::Action, TransactionKind), Error> {
     let tx_bytes = decode_b64(tx_bytes_b64)?;
     let tx_kind: EthTransactionKind = tx_bytes.as_slice().try_into()?;
     let tx: NormalizedEthTransaction = tx_kind.try_into()?;
-    let validation_outcome = validate_tx_relayer_data(&tx, target, context, *expected_nonce)?;
+    let target_kind = validate_tx_relayer_data(&tx, target, context, *expected_nonce)?;
 
     // If the transaction is valid then increment the nonce to prevent replay
     *expected_nonce = expected_nonce.saturating_add(1);
 
-    let to = tx.to.ok_or(Error::User(UserError::EvmDeployDisallowed))?.raw();
-    let action = if to != context.current_address
-        && extract_address(target).map(|a| a == to).unwrap_or(false)
-    {
-        // If target is another Ethereum implicit account then the action
-        // must be a transfer (because EOAs are not contracts on Ethereum).
-        Action::Transfer { receiver_id: target.to_string(), yocto_near: 0 }
-    } else {
-        parse_tx_data(target, &tx, context)?
+    // The way an honest relayer assigns `target` is as follows:
+    // 1. If the Ethereum transaction payload represents a Near action then use the receiver_id,
+    // 2. If the payload looks like a supported Ethereum emulation then use the address registrar:
+    // 2.a. if the tx.to address is registered then use the associated account id,
+    // 2.b. otherwise, tx.to == target
+    // 3. Otherwise, tx.to == target
+    // Given this algorithm, the only way to have `TargetKind::EthImplicit` is in the
+    // following cases:
+    // I)   The Ethereum transaction payload is not parseable as a known action,
+    // II)  The payload is parsable as a Near action and the receiver_id is an implicit account
```

Comment:
```suggestion
    // II)  The payload is parsable as a Near action and the receiver_id is an eth-implicit account
```

Otherwise if the receiver is a near-implicit account, I think it should be `TargetKind::OtherNearAccount`.

Reply:
Done in https://github.com/near/nearcore/pull/11435/commits/e5142cc526e870fffac11a2538667fa8a3565f9f